util/limiter/limiter.go

```diff
@@ -0,0 +1,124 @@
+package limiter
+
+import (
+	"sync"
+	"time"
+
+	"github.com/hashicorp/golang-lru/v2/expirable"
+)
+
+// Limiter is a keyed token bucket rate limiter.
+//
+// Each key gets its own separate token bucket to pull from, enabling
+// enforcement on things like "requests per IP address". To avoid
+// unbounded memory growth, Limiter actually only tracks limits
+// precisely for the N most recently seen keys, and assumes that
+// untracked keys are well-behaved. This trades off absolute precision
+// for bounded memory use, while still enforcing well for outlier
+// keys.
+//
+// As such, Limiter should only be used in situations where "rough"
+// enforcement of outliers only is sufficient, such as throttling
+// egregious outlier keys (e.g. something sending 100 queries per
+// second, where everyone else is sending at most 5).
+//
+// Each key's token bucket behaves like a regular token bucket, with
+// the added feature that a bucket's token count can optionally go
+// negative. This implements a form of "cooldown" for keys that exceed
+// the rate limit: once a key starts getting denied, it must stop
+// requesting tokens long enough for the bucket to return to a
+// positive balance. If the key keeps hammering the limiter in excess
+// of the rate limit, the token count will remain negative, and the
+// key will not be allowed to proceed at all. This is in contrast to
+// the classic token bucket, where a key trying to use more than the
+// rate limit will get capped at the limit, but can still occasionally
+// consume a token as one becomes available.
+type Limiter[K comparable] struct {
+	// Size is the number of keys to track. Only the Size most
+	// recently seen keys have their limits enforced precisely, older
+	// keys are assumed to not be querying frequently enough to bother
+	// tracking.
+	Size int
+	// TTL is an optional timeout for tracked keys. Keys not used for
```

Comment:
I'm not sure I understand how this works—is the idea that any key that was registered longer ago than this is dropped, and that the "recency" computation only applies to keys younger than that?

Reply:
Yes. In implementation terms: the LRU is a regular LRU of size N, but it periodically looks at the tail (least recently used) entries and evicts anything that's not been touched for longer than TTL.

So, if I set up a limiter tracking 1000 keys and with a 10-minute TTL, then hit it once with 1000 different keys, 10 minutes later the LRU will empty itself out because even though no other keys have showed up, all 1000 entries have exceeded their TTL without being touched.

Reply:
TBH this may be over-optimizing, the goal is to further limit memory overhead for a limiter that ends up only tracking a keyset that's smaller than Limiter.Size... But each entry only costs an extra 16 bytes, and the expected uses are for hundreds of keys not hundreds of thousands... So sticking to just regular LRU semantics is probably fine, on reflection. I'll switch to a non-expiring LRU and remove this knob.